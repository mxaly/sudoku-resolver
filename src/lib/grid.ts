export type sudokuValue = 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9;

export type cellAddress = [number, number];

interface CellData {
  value: sudokuValue | null;
  options: sudokuValue[];
  isConstant: boolean;
}

export interface Cell extends CellData {
  row: number;
  col: number;
  box: number;
}

export type Grid = Cell[];

const GRID_SIZE = 9;
export const OPTIONS = [...Array(9).keys()].map((k) => (k + 1) as sudokuValue);

export function createGrid(): Cell[] {
  const cells: Cell[] = [];
  for (let row = 0; row < GRID_SIZE; row++) {
    for (let col = 0; col < GRID_SIZE; col++) {
      cells.push(createCell({ row, col }));
    }
  }
  return cells;
}

export function createFromString(data: string): Cell[] {
  const grid = createGrid();
  data
    .replaceAll(/\s/g, "")
    .split("")
    .forEach((val, index) => {
      const row = Math.floor(index / 9);
      const col = index - row * 9;
      const num = parseInt(val);

      if (num > 0) {
        set(grid, [row, col], {
          value: num as sudokuValue,
          isConstant: true,
        });
      }
    });
  return grid;
}

function createCell({
  row,
  col,
  value,
  options,
  isConstant,
}: {
  row: number;
  col: number;
  value?: sudokuValue | null;
  options?: sudokuValue[];
  isConstant?: boolean;
}): Cell {
  return {
    value: value || null,
    options: options || [],
    row,
    col,
    box: getBoxIndex([row, col]),
    isConstant: !!isConstant,
  };
}

export function get(grid: Grid, addr: cellAddress): Cell {
  return grid[getIndex(addr)];
}

export function set(grid: Grid, addr: cellAddress, data: Partial<CellData>) {
  const cell = grid[getIndex(addr)];
  grid[getIndex(addr)] = { ...cell, ...data };
  return [...grid];
}

function getBoxIndex([rowIndex, colIndex]: cellAddress): number {
  const row = Math.floor(rowIndex / 3);
  const column = Math.floor(colIndex / 3);
  return row * 3 + column;
}

function getIndex([row, col]: [number, number]) {
  return row * GRID_SIZE + col;
}
