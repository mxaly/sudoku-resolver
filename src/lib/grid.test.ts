import { createGrid } from "./grid";
import { describe, it, expect } from "vitest";

describe("Grid", () => {
  test("createGrid", () => {
    expect(createGrid().cells).toMatchSnapshot();
  });

  test("set", () => {
    const grid = createGrid();

    grid.set([0, 0], { value: 3 });
    expect(grid.cells[0]).toEqual({
      box: 0,
      col: 0,
      options: [],
      row: 0,
      value: 3,
    });
  });

  test("get", () => {
    const grid = createGrid();

    expect(grid.get([8, 8])).toEqual({
      box: 8,
      col: 8,
      options: [],
      row: 8,
      value: null,
    });

    expect(grid.get([4, 6])).toEqual({
      box: 5,
      col: 6,
      options: [],
      row: 4,
      value: null,
    });
  });
});
