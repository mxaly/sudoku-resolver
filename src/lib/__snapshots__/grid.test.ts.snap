// Vitest Snapshot v1

exports[`Grid > createGrid 1`] = `
[
  {
    "box": 0,
    "col": 0,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 0,
    "col": 1,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 0,
    "col": 2,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 1,
    "col": 3,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 1,
    "col": 4,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 1,
    "col": 5,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 2,
    "col": 6,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 2,
    "col": 7,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 2,
    "col": 8,
    "options": [],
    "row": 0,
    "value": null,
  },
  {
    "box": 0,
    "col": 0,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 0,
    "col": 1,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 0,
    "col": 2,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 1,
    "col": 3,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 1,
    "col": 4,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 1,
    "col": 5,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 2,
    "col": 6,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 2,
    "col": 7,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 2,
    "col": 8,
    "options": [],
    "row": 1,
    "value": null,
  },
  {
    "box": 0,
    "col": 0,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 0,
    "col": 1,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 0,
    "col": 2,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 1,
    "col": 3,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 1,
    "col": 4,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 1,
    "col": 5,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 2,
    "col": 6,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 2,
    "col": 7,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 2,
    "col": 8,
    "options": [],
    "row": 2,
    "value": null,
  },
  {
    "box": 3,
    "col": 0,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 3,
    "col": 1,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 3,
    "col": 2,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 4,
    "col": 3,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 4,
    "col": 4,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 4,
    "col": 5,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 5,
    "col": 6,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 5,
    "col": 7,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 5,
    "col": 8,
    "options": [],
    "row": 3,
    "value": null,
  },
  {
    "box": 3,
    "col": 0,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 3,
    "col": 1,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 3,
    "col": 2,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 4,
    "col": 3,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 4,
    "col": 4,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 4,
    "col": 5,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 5,
    "col": 6,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 5,
    "col": 7,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 5,
    "col": 8,
    "options": [],
    "row": 4,
    "value": null,
  },
  {
    "box": 3,
    "col": 0,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 3,
    "col": 1,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 3,
    "col": 2,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 4,
    "col": 3,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 4,
    "col": 4,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 4,
    "col": 5,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 5,
    "col": 6,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 5,
    "col": 7,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 5,
    "col": 8,
    "options": [],
    "row": 5,
    "value": null,
  },
  {
    "box": 6,
    "col": 0,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 6,
    "col": 1,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 6,
    "col": 2,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 7,
    "col": 3,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 7,
    "col": 4,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 7,
    "col": 5,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 8,
    "col": 6,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 8,
    "col": 7,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 8,
    "col": 8,
    "options": [],
    "row": 6,
    "value": null,
  },
  {
    "box": 6,
    "col": 0,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 6,
    "col": 1,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 6,
    "col": 2,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 7,
    "col": 3,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 7,
    "col": 4,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 7,
    "col": 5,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 8,
    "col": 6,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 8,
    "col": 7,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 8,
    "col": 8,
    "options": [],
    "row": 7,
    "value": null,
  },
  {
    "box": 6,
    "col": 0,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 6,
    "col": 1,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 6,
    "col": 2,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 7,
    "col": 3,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 7,
    "col": 4,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 7,
    "col": 5,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 8,
    "col": 6,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 8,
    "col": 7,
    "options": [],
    "row": 8,
    "value": null,
  },
  {
    "box": 8,
    "col": 8,
    "options": [],
    "row": 8,
    "value": null,
  },
]
`;
